### PR TITLE
Reduce allocations

### PR DESCRIPTION
**Pull request by cdmihai**

### CODE REVIEW

**Comment by cdmihai** [comment]
### DesignTimeBuild Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 36.1077 -> 36.079 (-0.079%)
DotnetWebProject | :ok_hand: | no | 228.969 -> 229.701 (0.32%)
DotnetMvcProject | ::ok_hand: | no | 232.635 -> 230.3783 (-0.97%)
Picasso | :white_check_mark: | yes | 1564.0038 -> 1551.749 (-0.784%)
SmallP2POldCsproj | :red_circle: | yes | 67.6956 -> 67.9109 (0.318%)
SmallP2PNewCsproj | :red_circle: | yes | 650.8806 -> 654.7055 (0.588%)
LargeP2POldCsproj | :white_check_mark: | yes | 11367.3856 -> 11280.1764 (-0.767%)
OrchardCore | :red_circle: | yes | 48305.1718 -> 48473.7259 (0.349%)

### DesignTimeBuild Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 6262133 -> 6251411 (-0.171%)
DotnetWebProject | ::ok_hand: | no | 23643062 -> 23524839 (-0.5%)
DotnetMvcProject | :ok_hand: | no | 21273645 -> 21582133 (1.45%)
Picasso | ::ok_hand: | no | 185284419 -> 183769750 (-0.817%)
SmallP2POldCsproj | :white_check_mark: | yes | 8590609 -> 8525486 (-0.758%)
SmallP2PNewCsproj | ::ok_hand: | no | 91484792 -> 90769621 (-0.782%)
LargeP2POldCsproj | ::ok_hand: | no | 1144648931 -> 1139162127 (-0.479%)
OrchardCore | ::ok_hand: | no | 2712596216 -> 2655450741 (-2.107%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 35.0583 -> 34.6549 (-1.151%)
DotnetWebProject | :ok_hand: | no | 48.5349 -> 48.6267 (0.189%)
DotnetMvcProject | :ok_hand: | no | 54.3 -> 54.3388 (0.071%)
Picasso | :white_check_mark: | yes | 306.4513 -> 304.1176 (-0.762%)
SmallP2POldCsproj | :ok_hand: | no | 51.5858 -> 51.63 (0.086%)
SmallP2PNewCsproj | :ok_hand: | no | 207.9619 -> 208.1404 (0.086%)
LargeP2POldCsproj | :white_check_mark: | yes | 901.6358 -> 897.1365 (-0.499%)
OrchardCore | :ok_hand: | no | 3202.3458 -> 3210.1659 (0.244%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5685964 -> 5678588 (-0.13%)
DotnetWebProject | :white_check_mark: | yes | 7669798 -> 7652572 (-0.225%)
DotnetMvcProject | :white_check_mark: | yes | 8291518 -> 8274166 (-0.209%)
Picasso | :red_circle: | yes | 12007874 -> 13539754 (12.757%)
SmallP2POldCsproj | :white_check_mark: | yes | 7017402 -> 6989927 (-0.392%)
SmallP2PNewCsproj | :ok_hand: | no | 14087460 -> 14666999 (4.114%)
LargeP2POldCsproj | :ok_hand: | no | 39614507 -> 40180225 (1.428%)
OrchardCore | ::ok_hand: | no | 70787983 -> 70453100 (-0.473%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 29.5337 -> 29.5277 (-0.02%)
DotnetWebProject | ::ok_hand: | no | 36.8909 -> 36.8265 (-0.175%)
DotnetMvcProject | ::ok_hand: | no | 39.1836 -> 39.149 (-0.088%)
Picasso | ::ok_hand: | no | 235.3754 -> 235.2364 (-0.059%)
SmallP2POldCsproj | :white_check_mark: | yes | 41.7634 -> 41.3848 (-0.907%)
SmallP2PNewCsproj | :white_check_mark: | yes | 132.8121 -> 131.5824 (-0.926%)
LargeP2POldCsproj | :white_check_mark: | yes | 671.4277 -> 650.7794 (-3.075%)
Generated_100_100_v150 | :white_check_mark: | yes | 1145.0244 -> 1121.3286 (-2.069%)
OrchardCore | :red_circle: | yes | 1566.2614 -> 1578.4338 (0.777%)
Roslyn | :white_check_mark: | yes | 2697.9207 -> 2677.5382 (-0.755%)
WebLargeCore | :white_check_mark: | yes | 1810.9627 -> 1790.5784 (-1.126%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5550925 -> 5542880 (-0.145%)
DotnetWebProject | :white_check_mark: | yes | 6866739 -> 6861580 (-0.075%)
DotnetMvcProject | :white_check_mark: | yes | 7300617 -> 7288526 (-0.166%)
Picasso | ::ok_hand: | no | 12016873 -> 11041749 (-8.115%)
SmallP2POldCsproj | :white_check_mark: | yes | 6960152 -> 6935741 (-0.351%)
SmallP2PNewCsproj | :white_check_mark: | yes | 23115184 -> 23032763 (-0.357%)
LargeP2POldCsproj | :ok_hand: | no | 36616242 -> 37150478 (1.459%)
Generated_100_100_v150 | :white_check_mark: | yes | 46183934 -> 45110769 (-2.324%)
OrchardCore | :white_check_mark: | yes | 60028228 -> 57018526 (-5.014%)
Roslyn | ::ok_hand: | no | 95657140 -> 94602436 (-1.103%)
WebLargeCore | ::ok_hand: | no | 71829902 -> 71151187 (-0.945%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 33.8655 -> 34.0576 (0.567%)
DotnetWebProject | ::ok_hand: | no | 47.5824 -> 47.5323 (-0.105%)
DotnetMvcProject | :ok_hand: | no | 53.1515 -> 53.1802 (0.054%)
Picasso | :white_check_mark: | yes | 250.8027 -> 233.6591 (-6.835%)
SmallP2POldCsproj | :white_check_mark: | yes | 45.9033 -> 45.3374 (-1.233%)
SmallP2PNewCsproj | ::ok_hand: | no | 170.848 -> 170.4664 (-0.223%)
LargeP2POldCsproj | :white_check_mark: | yes | 728.2145 -> 715.376 (-1.763%)
Generated_100_100_v150 | :white_check_mark: | yes | 1160.7843 -> 1141.1447 (-1.692%)
OrchardCore | :red_circle: | yes | 2114.6526 -> 2123.8197 (0.434%)
Roslyn | :white_check_mark: | yes | 3228.1026 -> 3195.6526 (-1.005%)
WebLargeCore | :white_check_mark: | yes | 2288.9533 -> 2254.6023 (-1.501%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.master vs src\msbuild.mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5755935 -> 5747848 (-0.14%)
DotnetWebProject | :white_check_mark: | yes | 7726999 -> 7706464 (-0.266%)
DotnetMvcProject | :white_check_mark: | yes | 8329708 -> 8324415 (-0.064%)
Picasso | :red_circle: | yes | 11244682 -> 36661979 (226.038%)
SmallP2POldCsproj | :white_check_mark: | yes | 7025170 -> 6992965 (-0.458%)
SmallP2PNewCsproj | :ok_hand: | no | 13001039 -> 13700658 (5.381%)
LargeP2POldCsproj | :white_check_mark: | yes | 38329360 -> 37686868 (-1.676%)
Generated_100_100_v150 | :white_check_mark: | yes | 46640236 -> 45856827 (-1.68%)
OrchardCore | :ok_hand: | no | 64107643 -> 68600667 (7.009%)
Roslyn | ::ok_hand: | no | 92480710 -> 90373331 (-2.279%)
WebLargeCore | ::ok_hand: | no | 66576846 -> 66408762 (-0.252%)

File: src/Build/BackEnd/Components/RequestBuilder/TargetUpToDateChecker.cs [hunk 1]
```diff
@@ -351,8 +352,31 @@ private static string GetIncrementalBuildReason(DependencyAnalysisLogDetail logD
         /// </summary>
         private void LogUniqueInputsAndOutputs()
         {
-            _loggingService.LogComment(_buildEventContext, MessageImportance.Low, "SkipTargetUpToDateInputs", string.Join(";", _uniqueTargetInputs.Keys));
-            _loggingService.LogComment(_buildEventContext, MessageImportance.Low, "SkipTargetUpToDateOutputs", string.Join(";", _uniqueTargetOutputs.Keys));
+            var targetInputKeys = _uniqueTargetInputs.Keys;
+            var targetOutputKeys = _uniqueTargetOutputs.Keys;
+
+            var maxLengthSum = Math.Max(LengthSum(targetInputKeys), LengthSum(targetOutputKeys));
+            var maxLength = Math.Max(targetInputKeys.Count, targetOutputKeys.Count);
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TargetUpToDateChecker.cs]: name this `joinerCount` or `semicolonCount` or something? I thought the length calculation looked fishy because the name confused me.

File: src/Shared/ReuseableStringBuilder.cs [hunk 2]
```diff
@@ -186,6 +186,35 @@ internal ReuseableStringBuilder Append(string value, int startIndex, int count)
             return this;
         }
 
+        public ReuseableStringBuilder AppendJoined(char separator, ICollection<string> strings)
```
Comment (rainersigwald) [inline src/Shared/ReuseableStringBuilder.cs]: `AppendSeparated`? I like the consistency with `Join` but the name feels wrong to me.

File: src/Shared/ReuseableStringBuilder.cs [hunk 3]
```diff
@@ -186,6 +186,35 @@ internal ReuseableStringBuilder Append(string value, int startIndex, int count)
...
+            _cachedString = null;
+
+            var i = 0;
+
+            foreach (var s in strings)
+            {
+                _borrowedBuilder.Append(s);
+
+                i++;
+                if (i < strings.Count - 1)
```
Comment (rainersigwald) [inline src/Shared/ReuseableStringBuilder.cs]: extreme nitpick: Maybe you could count _down_ from `int separatorsRemaining = strings.Count - 1` to avoid having to get the count multiple times and simplify the condition?

File: src/Tasks/Hash.cs [hunk 4]
```diff
@@ -37,28 +38,42 @@ public override bool Execute()
...
             {
-                StringBuilder hashInput = new StringBuilder();
-
-                foreach (var item in ItemsToHash)
-                {
-                    hashInput.Append(item.ItemSpec);
-                    hashInput.Append(ItemSeparatorCharacter);
-                }
-
                 using (var sha1 = SHA1.Create())
```
Comment (rainersigwald) [inline src/Tasks/Hash.cs]: Did you move the SHA using up for a reason? It looks like you could put it inside the stringbuilder borrow.

File: src/Tasks/Hash.cs [hunk 5]
```diff
@@ -37,28 +38,42 @@ public override bool Execute()
...
+                        {
+                            stringBuilder.Append(item.ItemSpec);
+                            stringBuilder.Append(ItemSeparatorCharacter);
+                        }
+
+                        //todo return the string builder bytes when the APIs will get that overload
+                        hash = sha1.ComputeHash(Encoding.UTF8.GetBytes(stringBuilder.ToString()));
+                    }
+
+                    using (var hashResult = new ReuseableStringBuilder(hash.Length * 2))
```
Comment (rainersigwald) [inline src/Tasks/Hash.cs]: Think it's worth including this in the length computation above to use a single stringbuilder for this whole operation?

File: src/Tasks/Hash.cs [hunk 6]
```diff
@@ -37,28 +38,42 @@ public override bool Execute()
...
-                    HashResult = hashResult.ToString();
+                    using (var stringBuilder = new ReuseableStringBuilder(totalItemSize + ItemsToHash.Length))
+                    {
+                        foreach (var item in ItemsToHash)
+                        {
+                            stringBuilder.Append(item.ItemSpec);
+                            stringBuilder.Append(ItemSeparatorCharacter);
+                        }
+
+                        //todo return the string builder bytes when the APIs will get that overload
```
Comment (rainersigwald) [inline src/Tasks/Hash.cs]: Not sure I understand this comment. Link to a corefx issue or something?

File: src/Shared/ResourceUtilities.cs [hunk 7]
```diff
@@ -190,14 +190,27 @@ internal static string FormatResourceString(string resourceName)
         /// <param name="resourceName">Resource string to load.</param>
         /// <param name="args">Optional arguments for formatting the resource string.</param>
         /// <returns>The formatted resource string.</returns>
-        internal static string FormatResourceString(string resourceName, params object[] args)
+        internal static string FormatResourceStringStripCodeAndKeyWord(string resourceName, params object[] args)
```
Comment (rainersigwald) [inline src/Shared/ResourceUtilities.cs]: Can you do this rename in a separate commit with a comment that it's a pure rename with no semantic change? This looked scary but actually wasn't!

File: src/Shared/ResourceUtilities.cs [hunk 8]
```diff
@@ -190,14 +190,27 @@ internal static string FormatResourceString(string resourceName)
...
         }
 
+        /// <summary>
+        /// Formats the resource string with the given arguments.
+        /// Ignores error codes and keywords
+        /// </summary>
+        /// <param name="resourceName"></param>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        internal static string FormatResourceStringIgnoreCode(string resourceName, params object[] args)
```
Comment (rainersigwald) [inline src/Shared/ResourceUtilities.cs]: Should this be parallel in naming to the above, which says `Strip` instead of `Ignore`?

Comment (cdmihai) [reply src/Shared/ResourceUtilities.cs]: If only there was some git extension / IDE interaction that provided better support for refactorings, as in separating them in extra commits, or at least adding some metadata to the diffs to tie them together and give them a higher level name.

**Comment by cdmihai** [comment]
@rainersigwald implemented the feedback but squashed the changes into the original commits because the renames would have made it more difficult otherwise :)
